src/Testable/Navigation.elm

```diff
@@ -0,0 +1,156 @@
+module Testable.Navigation exposing (..)
+
+import List.Extra
+import
+    -- This "unused" import is required for the native bindings
+    Native.Testable.Navigation
+import Navigation exposing (Location)
+import Regex exposing (HowMany(..), find, regex)
+
+
+type alias History =
+    ( Int, List Location )
+
+
+type Msg
+    = Jump Int
+    | New String
+    | Modify String
+    | Visit String
+
+
+type ReturnMsg
+    = NoOp
+    | ReturnLocation Location
+    | TriggerLocationMsg Location
+
+
+currentLocation : History -> Location
+currentLocation ( index, history ) =
+    List.Extra.getAt index history
+        |> Maybe.withDefault (getLocation "")
+
+
+update : Msg -> History -> ( History, ReturnMsg )
+update msg ( index, history ) =
+    let
+        location =
+            currentLocation ( index, history )
+    in
+    case msg of
+        Jump n ->
+            let
+                nextIndex =
+                    min (max (index + n) 0) (List.length history - 1)
+
+                nextLocation =
+                    List.Extra.getAt nextIndex history
+                        |> Maybe.withDefault (getLocation "")
+            in
+            ( ( nextIndex, history ), TriggerLocationMsg nextLocation )
+
+        New url ->
+            let
+                nextLocation =
+                    setLocation url location
+
+                modifiedHistory =
+                    List.take (index + 1) history ++ [ nextLocation ]
+            in
+            ( ( index + 1, modifiedHistory ), ReturnLocation nextLocation )
+
+        Modify url ->
+            let
+                nextLocation =
+                    setLocation url location
+
+                modifiedHistory =
+                    history
+                        |> List.Extra.updateIfIndex ((==) index) (always nextLocation)
+
+                effect =
+                    if nextLocation == location then
+                        NoOp
+                    else
+                        ReturnLocation nextLocation
+            in
+            ( ( index, modifiedHistory ), effect )
+
+        Visit url ->
+            let
+                ( modifiedHistory, _ ) =
+                    update (New url) ( index, history )
+            in
+            ( modifiedHistory, TriggerLocationMsg (currentLocation modifiedHistory) )
+
+
+init : History
+init =
+    ( 0, [ initialLocation ] )
+
+
+initialLocation : Location
+initialLocation =
+    getLocation "https://elm.testable/"
```

Comment:
hmm.. If you use `Navigation.back` with a regular program, won't it always leave the program (or is there some way to intercept browser navigation without using `Navigation.program`?)

Even though both `Html.program` and `Navigation.program` return the same Elm types, we can still store hidden information in the javascript object so we can tell them apart at runtime.  This is how things currently work with flags (even though in the case of flags, the compiler knows something at the type level, there is a runtime check and some error handling logic that is performed at runtime).